source/h2_connection.c

```diff
@@ -0,0 +1,145 @@
+    .on_request = NULL,
+    .on_response = NULL,
+    .on_header = NULL,
+    .on_body = NULL,
+    .on_done = NULL,
+};
+
+/* Common new() logic for server & client */
+static struct aws_h2_connection *s_connection_new(
+    struct aws_allocator *alloc,
+    size_t initial_window_size,
+    bool server) {
+
+    (void)server;
+
+    struct aws_h2_connection *connection = aws_mem_calloc(alloc, 1, sizeof(struct aws_h2_connection));
+    if (!connection) {
+        goto error_connection_alloc;
+    }
+
+    connection->base.vtable = &s_h2_connection_vtable;
+    connection->base.alloc = alloc;
+    connection->base.channel_handler.vtable = &s_h2_connection_vtable.channel_handler_vtable;
+    connection->base.channel_handler.impl = connection;
+    connection->base.http_version = AWS_HTTP_VERSION_1_1;
+    connection->base.initial_window_size = initial_window_size;
+
+    /* 1 refcount for user */
+    aws_atomic_init_int(&connection->base.refcount, 1);
+
+    /* Init the next stream id (server must use odd ids, client even [RFC 7540 5.1.1])*/
+    aws_atomic_init_int(&connection->stream_id, server ? 2 : 1);
+
+    /* Create a new decoder */
+    struct aws_h2_decoder_params params = {
+        .alloc = alloc,
+        .user_data = connection,
+        .vtable = s_h2_decoder_vtable,
+    };
+    connection->thread_data.decoder = aws_h2_decoder_new(&params);
+
+    int err = aws_mutex_init(&connection->synced_data.lock);
+    if (err) {
+        CONNECTION_LOGF(
+            ERROR, connection, "static: Failed to initialize mutex, error %d (%s).", err, aws_error_name(err));
+
+        goto error_mutex;
+    }
+
+    return connection;
+
+error_mutex:
+    aws_mem_release(alloc, connection);
+error_connection_alloc:
+    return NULL;
+}
+
+struct aws_http_connection *aws_http_connection_new_http2_server(
+    struct aws_allocator *allocator,
+    size_t initial_window_size) {
+
+    struct aws_h2_connection *connection = s_connection_new(allocator, initial_window_size, true);
+    if (!connection) {
+        return NULL;
+    }
+
+    connection->base.server_data = &connection->base.client_or_server_data.server;
+
+    return &connection->base;
+}
+
+struct aws_http_connection *aws_http_connection_new_http2_client(
+    struct aws_allocator *allocator,
+    size_t initial_window_size) {
+
+    struct aws_h2_connection *connection = s_connection_new(allocator, initial_window_size, false);
+    if (!connection) {
+        return NULL;
+    }
+
+    connection->base.client_data = &connection->base.client_or_server_data.client;
+
+    return &connection->base;
+}
+
+uint32_t aws_h2_connection_get_next_stream_id(struct aws_h2_connection *connection) {
+    size_t next_id = aws_atomic_fetch_add(&connection->stream_id, 2);
+
+    /* Check for next_id going over UINT32_MAX or overflowing when size_t is 32 bits */
+    if (AWS_UNLIKELY(next_id > UINT32_MAX || (next_id - 2) > next_id)) {
+        aws_raise_error(AWS_ERROR_HTTP_PROTOCOL_ERROR);
```

Comment:
Edge case: multiple threads try and create a new request simultaneously and we're about to overflow: only the first will detect the overflow.

Maybe don't do this as an atomic? We're inevitable going to need to grab a lock at some point while creating a thread to jam it into some linked-list in `synced_data`.